server/src/main/java/io/crate/analyze/expressions/ExpressionAnalyzer.java

```diff
@@ -1162,7 +1163,7 @@ private static void verifyTypesForMatch(Iterable<? extends Symbol> columns, Data
     }
 
     private static void ensureResultTypesMatch(Collection<? extends Symbol> results) {
-        HashSet<DataType> resultTypes = new HashSet<>();
+        HashSet<DataType<?>> resultTypes = new LinkedHashSet<>(results.size());
```

Comment:
Maybe instead of switching to a `LinkedHashSet` we could sort the types inside the `throws` block. That we we'd only pay the price in the error case instead of all the time.